src/lesson_4/FractionalWrapper.sol

```diff
@@ -0,0 +1,170 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.12;
+
+import {ERC20} from "yield-utils-v2/contracts/token/ERC20.sol";
+import {IERC20} from "yield-utils-v2/contracts/token/IERC20.sol";
+import {TransferHelper} from "yield-utils-v2/contracts/token/TransferHelper.sol";
+
+/**
+@title An Fractional ERC20 wrapper/vault
+@author Sabnock01
+@notice You can use this contract to wrap and unwrap a specified ERC20 token on a fractionalized basis
+@dev The ERC20 token is set in the constructor and the wrapped tokens
+are burned upon unwrapping
+*/
+contract FractionalWrapper is ERC20 {
+    using TransferHelper for IERC20;
+    
+    ///@notice The ERC20 token to wrap
+    ///@dev technically a type address to be EIP-4626 compliant but IERC20 is recognized as such by the compiler
+    IERC20 public asset;
+    ///@notice exchange rate between asset and underlying
+    ///@dev set to 26 decimal places
+    uint256 public exchangeRate;
+
+    ///@notice Event emitted when tokens are wrapped
+    event Deposit(
+        address indexed caller, 
+        address indexed owner, 
+        uint256 assets, 
+        uint256 shares
+    );
+
+    ///@notice Event emitted when tokens are unwrapped
+    event Withdraw(
+        address indexed caller, 
+        address indexed receiver, 
+        address indexed owner, 
+        uint256 assets, 
+        uint256 shares
+    );
+
+    ///@notice Creates a fractionalized wrapper for a given ERC20 token
+    ///@param token the underlying asset
+    ///@param name the name of the underlying
+    ///@param symbol the symbol of the underlying
+    constructor(IERC20 token, uint256 _exchangeRate, string memory name, string memory symbol) ERC20(name, symbol, 18) {
+        asset = token;
+        exchangeRate = _exchangeRate;
+    }
+
+    ///@notice Total amount of the underlying asset that is “managed” by the vault
+    ///@return totalManagedAssets balance of the underlying token within the vault
+    function totalAssets() public view returns (uint256 totalManagedAssets) {
+        return asset.balanceOf(address(this));
+    }
+
+    ///@notice Calculates the amount of the asset (wrapped) token the user can get for their amount of the underlying
+    ///@param assets amount of the underlying token EX. DAI
+    ///@return shares amount of the asset (wrapped) token EX. fyDAI
+    function convertToShares(uint256 assets) public view returns (uint256 shares) {
+        return (assets * exchangeRate) / 1e27;
```

Comment:
You call `convertToShares` several times from other functions in this contract. Every time that happens, since `convertToShares` is `public`, your contract does a CALL to itself.

This CALL changes `msg.sender`, which is of no consequence. This CALL costs more gas than calling an internal function (which is a JUMP). I'm not sure how much extra gas this is since addresses are cached. It could be 2000, or it could be 100. Still, it's more expensive than a JUMP.

The usual pattern when you find that you need to call other functions in the same contract is to split them into an `external` function (`convertToShares`) that might do some verification, and an internal function (`_convertToShares`) that has the logic. The external calls the internal function so for external users it's basically the same. Other functions in your contract (`previewDeposit`, for example) would call the internal function (`_convertToShares`) and save gas.

If you check almost any contract in the yieldprotocol account you will find examples of this pattern.